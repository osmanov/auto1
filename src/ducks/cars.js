import { createSelector } from 'reselect'
import { all, call, put, take } from 'redux-saga/effects'
import { appName, apiURL } from '../config'
import request from '../utils/request'

export const moduleName = 'cars'
const prefix = `${appName}/${moduleName}`

export const FETCH_CARS_REQUEST = `${prefix}/FETCH_CARS_REQUEST`
export const FETCH_CARS_SUCCESS = `${prefix}/FETCH_CARS_SUCCESS`
export const FETCH_CARS_ERROR = `${prefix}/FETCH_CARS_ERROR`

export const initialState = {
  loading: false,
  data: [],
  error: null,
  sort: null,
  filter: null,
  totalPageCount: 0
}

export const stateSelector = state => state[moduleName]
export const filterSelector = state => state[moduleName].filter
export const totalPageCountSelector = state => state[moduleName].totalPageCount
export const sortSelector = state => state[moduleName].sort

export const currentPageSelector = createSelector(
  filterSelector,
  filter => {
    return filter ? filter.page : 1
  }
)

export const paginatorParamsSelector = createSelector(
  currentPageSelector,
  totalPageCountSelector,
  filterSelector,
  sortSelector,
  (current, totalCount, filter, sort) => {
    const nextCandidate = current + 1
    const prevCandidate = current - 1

    return {
      next: totalCount >= nextCandidate ? nextCandidate : null,
      prev: prevCandidate || null,
      current,
      first: 1,
      last: totalCount,
      sort,
      filter: filter || {}
    }
  }
)
export const buildURL = params => {
  return Object.keys(params).reduce((acc, param) => {
    return params[param] ? `${acc}&${param}=${params[param]}` : acc
  }, `${apiURL}/cars?`)
}

export default function reducer(state = initialState, action) {
  const { type, payload, error } = action
  switch (type) {
    case FETCH_CARS_REQUEST:
      const { sort, ...filter } = payload
      return { ...state, loading: true, sort, filter }
    case FETCH_CARS_SUCCESS:
      const { cars, ...other } = payload
      return {
        ...state,
        ...other,
        loading: false,
        data: payload.cars,
        error: null
      }
    case FETCH_CARS_ERROR:
      return { ...state, loading: false, error }
    default:
      return state
  }
}

export function fetchCars({ sort, page = 1, manufacturer, color } = {}) {
  return {
    type: FETCH_CARS_REQUEST,
    payload: { sort, page, manufacturer, color }
  }
}
export const fetchCarsSaga = function*() {
  while (true) {
    const { payload } = yield take(FETCH_CARS_REQUEST)
    const requestURL = Object.keys(payload).reduce((acc, param) => {
      return payload[param] ? `${acc}&${param}=${payload[param]}` : acc
    }, `${apiURL}/cars?`)

    try {
      const data = yield call(request, requestURL)
      yield put({
        type: FETCH_CARS_SUCCESS,
        payload: data
      })
    } catch (error) {
      yield put({
        type: FETCH_CARS_ERROR,
        error
      })
    }
  }
}

export const saga = function*() {
  yield all([fetchCarsSaga()])
}
