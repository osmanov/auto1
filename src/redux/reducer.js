import { combineReducers } from 'redux'
import carsReducer, { moduleName as carsModule } from '../ducks/cars'
import carReducer, { moduleName as carModule } from '../ducks/car'
import propertiesReducer, {
  moduleName as propertiesModule
} from '../ducks/properties'

export default combineReducers({
  [carsModule]: carsReducer,
  [carModule]: carReducer,
  [propertiesModule]: propertiesReducer
})
