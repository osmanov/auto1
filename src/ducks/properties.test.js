import { put } from 'redux-saga/effects'
import reducer, {
  initialState,
  fetchPropertiesSaga,
  FETCH_PROPERTIES_SUCCESS,
  FETCH_PROPERTIES_ERROR
} from './properties'

import { fetchCars } from './cars'

const mock = {
  colors: ['red', 'blue', 'green'],
  manufacturers: [{ models: { name: 'Avenger' }, name: 'Dodge' }]
}
const state = Object.freeze(initialState)
/**
 * Reducer Tests
 */
it('reducer should get properties', () => {
  const newState = reducer(state, {
    type: FETCH_PROPERTIES_SUCCESS,
    payload: mock
  })
  expect(newState).toMatchSnapshot()
})
/**
 * Saga Tests
 */
it('saga should get properties and start fetch cars', () => {
  const saga = fetchPropertiesSaga()
  expect(saga.next().value).toMatchSnapshot()
  expect(saga.next(mock).value).toMatchSnapshot()
  const data = [mock.colors, mock.manufacturers]
  const allDescriptor = saga.next([data]).value
  expect(allDescriptor).toMatchSnapshot()
  expect(saga.next().value).toEqual(put(fetchCars()))
})

it('saga should not get properties', () => {
  const saga = fetchPropertiesSaga()
  expect(saga.next().value).toMatchSnapshot()
  expect(saga.next(mock).value).toMatchSnapshot()
  const error = new Error()
  expect(saga.throw(error).value).toEqual(
    put({
      type: FETCH_PROPERTIES_ERROR,
      error
    })
  )
})
