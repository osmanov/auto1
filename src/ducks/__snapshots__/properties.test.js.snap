// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer should get properties 1`] = `
Object {
  "data": Object {
    "colors": Array [
      "red",
      "blue",
      "green",
    ],
    "manufacturers": Array [
      Object {
        "models": Object {
          "name": "Avenger",
        },
        "name": "Dodge",
      },
    ],
  },
  "error": null,
  "loaded": true,
  "loading": false,
}
`;

exports[`saga should get properties and start fetch cars 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "pattern": "auto1/properties/FETCH_PROPERTIES_REQUEST",
  },
  "type": "TAKE",
}
`;

exports[`saga should get properties and start fetch cars 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": true,
  "payload": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "http://localhost:3001/colors",
        ],
        "context": null,
        "fn": [Function],
      },
      "type": "CALL",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "http://localhost:3001/manufacturers",
        ],
        "context": null,
        "fn": [Function],
      },
      "type": "CALL",
    },
  ],
  "type": "ALL",
}
`;

exports[`saga should get properties and start fetch cars 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "payload": Object {
        "0": Array [
          "red",
          "blue",
          "green",
        ],
        "1": Array [
          Object {
            "models": Object {
              "name": "Avenger",
            },
            "name": "Dodge",
          },
        ],
      },
      "type": "auto1/properties/FETCH_PROPERTIES_SUCCESS",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`saga should not get properties 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "pattern": "auto1/properties/FETCH_PROPERTIES_REQUEST",
  },
  "type": "TAKE",
}
`;

exports[`saga should not get properties 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": true,
  "payload": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "http://localhost:3001/colors",
        ],
        "context": null,
        "fn": [Function],
      },
      "type": "CALL",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "http://localhost:3001/manufacturers",
        ],
        "context": null,
        "fn": [Function],
      },
      "type": "CALL",
    },
  ],
  "type": "ALL",
}
`;
